tests/interfaces/test_jax.py

```diff
@@ -146,20 +146,21 @@ def cost(params):
 
         # adjoint method only performs a single device execution per tape, but gets both result and gradient
         assert dev.num_executions == 3
-        spy.assert_called()
+        spy.assert_not_called()
```

Comment:
what caused this change?

Reply:
It's often more efficient to calculate the jacobian at same time as the results (`grad_on_execution=True`). But what if we aren't actually needing the derivatives? Then we did a bunch of work that never gets used.

In an ideal world, we would distinguish between "I just need the results" and "I'm doing a gradient calculation in the near future; please calculate the jacobians too".  

With jax, we can distinguish between those two cases. In `cost(x)`, we will use `_execute_wrapper`, and in `jax.grad(cost)(x)`, we will use `_execute_and_compute_jvp`. Two separate pathways.

Leads to some pretty nice speedups when we won't be needing the derivatives.

![Screenshot 2023-11-29 at 4 32 57 PM](https://github.com/PennyLaneAI/pennylane/assets/6364575/783ddf43-6ff9-43b6-b45b-faefe2b7ca2b)